main/resource-access.bs

```diff
@@ -70,7 +70,99 @@ A Solid data pod MUST conform to the Web Access Control specification [[!WAC]].
 
 ## Cross-Origin Resource Sharing ## {#cors}
 
-Issue: Write Cross-Origin Resource Sharing section.
-
-Draft:
-A Solid data pod MUST conform to the CORS specification [[!FETCH]].
+### Background and Need ### {#cors-need}
+<em>This section is non-normative.</em>
+
+[=Solid apps=] typically access data from multiple sources.
+However,
+Web browsers by default prevent apps that run on one origin
+from accessing data on other origins.
+This cross-origin protection is a security mechanism
+that ensures malicious websites cannot simply read
+your profile or banking details from other websites.
+However, this reasonable default poses a problem
+even for benevolent Solid apps,
+which might have good reasons to access data from different places.
+For instance,
+a Solid app at `https://app.example/`
+would be prevented from accessing data on
+`https://alice-data-pod.example/` or `https://bob-data-pod.example/`,
+even when Alice and Bob have given the user of the app
+their permission to see some of their data.
+
+For cases where the other origins
+have their own access protection mechanism—
+[like within Solid](#wac)—
+the browser's built-in cross-origin protection
+is actually an obstacle rather than a feature.
+After all,
+[=data pods=] already ensure through access control
+that certain documents can only be accessed by specific people.
+Preventively blocking apps from different origins
+thus introduces an unnecessary barrier.
+
+Fortunately,
+Web servers can indicate to the browser
+that certain documents do not require cross-origin protection.
+This mechanism to selectively disable that protection
+is called *Cross-Origin Resource Sharing* or *CORS* [[FETCH]].
+By responding to browser requests
+with a specific combination of HTTP headers,
+servers can indicate which actions are allowed for a given resource.
+For a Solid data pod,
+the goal is to allow *all* actions on the CORS level,
+such that the deeper [access control layer](#wac)
+can exert full control over the app's allowed permissions.
+The next section describes how to achieve this
+through the right HTTP header configuration.
+
+
+### Required server-side implementation ### {#cors-server}
+
+A [=data pod=] MUST implement the CORS protocol [[!FETCH]]
+such that, to the extent possible,
+the browser does not block
+any request sent to it or response received from it.
+If the data pod wishes to block access to a resource,
+this MUST NOT happen via CORS
+but MUST instead be communicated to the Solid app in the browser
+through HTTP status codes such as
+`401`, `403`, or `404` [[!RFC7231]].
+Furthermore,
+all response headers MUST be exposed to the Solid app.
```

Comment:
all with the exception of identifiable information and possibly something else?

Reply:
I don't think there are any exceptions. That is, at the moment identifiable information is transferred, the server has established the identity of the user.

If there were exceptions, then these could be bent by having a command-line client _pretend_ to be a browser (or forward a request on behalf of a browser), and that way stealing the information anyway.

The crucial notion is that we do _not_ rely on the browser's same-origin protection mechanisms for security, but rather on the authentication and authorization mechanism that Solid implements. So anything the server sends has been guaranteed to be safe by another mechanism.

Do you have concrete examples of headers? Or would you suggest to make this a SHOULD?

Reply:
I'd like to understand how to qualify what's in the set of "all". What breaks if not "all"?

Reply:
In general, HTTP headers convey useful or important information to the client. `Content-Type` for instance, but also more rare headers like `Link`, or even custom ones like `WAC-Allow`. What they all have in common is that, one way or another, the server wants to tell the client something about the enclosed representation.

And in regular Web apps, where the script requesting the representation is hosted on the same origin as the representation, the script can see all headers.

However, when origins are different—as is the case with Solid—access to several headers is blocked because, if the server relies on cross-origin protection as an authentication mechanism, such headers are sensitive.

Yet Solid servers do not rely on such an authentication mechanism.

Hence, we are requesting seeing all headers like a regular Web app would.

So what breaks if not all: in general, Solid apps would have fewer capabilities than regular Web apps (despite resources being authenticated).

Reply:
I understand that. I had a different interpretation of the text because I didn't catch the emphasis/context of "expose".  Perhaps we can lift that a bit by mentioning `ACAH`?

Reply:
👌